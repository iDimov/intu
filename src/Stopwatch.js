import React, { Component } from "react";

class Stopwatch extends Component {
  state = {
    running: false,
    TimeDifference: 0,
    lastTick: 0,
    laps: []
  };

  componentDidMount() {
    this.interval = setInterval(this.tick, 10);
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  handleStartStop = () => {
    if (this.state.running) {
      this.setState({
        running: false
      });
    } else {
      this.setState({
        running: true,
        lastTick: Date.now()
      });
    }
  };

  handleReset = () => {
    this.setState({
      running: false,
      TimeDifference: 0,
      lastTick: 0
    });
  };

  handleLap = () => {
    if (this.state.TimeDifference !== 0) {
      if (this.state.laps[0] !== this.state.TimeDifference) {
        this.setState({
          laps: [this.state.TimeDifference, ...this.state.laps]
        });
      }
    }
  };

  handleClearLaps = () => {
    this.setState({
      laps: []
    });
  };

  tick = () => {
    if (this.state.running) {
      let now = Date.now();
      let TimeDifference = now - this.state.lastTick;

      this.setState({
        TimeDifference: this.state.TimeDifference + TimeDifference,
        lastTick: now
      });
    }
  };

  format = milliseconds => {
    let totalSeconds = Math.floor(milliseconds / 1000);
    let minutes = Math.floor(totalSeconds / 60);
    let seconds = totalSeconds % 60;
    let totalMilliseconds = milliseconds % 1000;
    let formatedMS;

    if (totalMilliseconds > 9) {
      if (totalMilliseconds > 99) {
        formatedMS = `${totalMilliseconds}`;
      } else {
        formatedMS = `0${totalMilliseconds}`;
      }
    } else {
      formatedMS = `00${totalMilliseconds}`;
    }

    return `${minutes > 9 ? minutes : "0" + minutes} : ${
      seconds > 9 ? seconds : "0" + seconds
    } : ${formatedMS}`;
  };

  render() {
    let stopwatch = this.format(this.state.TimeDifference);
    let laps = this.state.laps.map(v => <li key={v}> {this.format(v)}</li>);
    return (
      <React.Fragment>
        <div className="Controls">
          <button href="#" className="start" onClick={this.handleStartStop}>
            {this.state.running ? "Stop" : "Start"}
          </button>
          <button href="#" onClick={this.handleLap}>
            Lap
          </button>
          <button href="#" onClick={this.handleReset}>
            Reset
          </button>
          <button href="#" onClick={this.handleClearLaps}>
            Clear Laps
          </button>
        </div>
        <div className="Stopwatch">{stopwatch}</div>
        <ul className="Results">{laps}</ul>
      </React.Fragment>
    );
  }
}

export default Stopwatch;
