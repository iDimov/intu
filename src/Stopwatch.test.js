import React from "react";
import ReactDOM from "react-dom";
import Stopwatch from "./Stopwatch";
import { configure } from "enzyme";
import { shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import renderer from "react-test-renderer";

configure({ adapter: new Adapter() });

const wrapper = shallow(<Stopwatch />);

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<Stopwatch />, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe("Stopwatch Component:", () => {
  describe("should include basic JSX:", () => {
    describe("should matches the snapshot", () => {
      const tree = renderer.create(<Stopwatch />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("should have an initial state:", () => {
    describe("running — boolean property, describing the state of stopwatch", () => {
      test("property should exist", () => {
        expect(wrapper.state("running")).toBeDefined();
      });
      test("should have false as the initial value", () => {
        expect(wrapper.state("running")).toBe(false);
      });
    });
    describe("TimeDifference — time difference between now and last call of tick function", () => {
      test("property should exist", () => {
        expect(wrapper.state("TimeDifference")).toBeDefined();
      });
      test("should have 0 as the initial value", () => {
        expect(wrapper.state("TimeDifference")).toBe(0);
      });
    });
    describe("lastTick — last value of tick function (Date.now())", () => {
      test("property should exist", () => {
        expect(wrapper.state("lastTick")).toBeDefined();
      });
      test("should have 0 as the initial value", () => {
        expect(wrapper.state("lastTick")).toBe(0);
      });
    });
  });

  describe("should contain class methods implemented through the arrow methods (class properties)", () => {
    afterAll(() => {
      jest.clearAllMocks();
    });
    describe("format method", () => {
      test("should calculate values from a given number of milliseconds and return a string in the format '00 : 00 : 000'", () => {
        expect(wrapper.instance().format(2742)).toBe("00 : 02 : 742");
      });
    });
    describe("handleStartStop", () => {
      test("handleStartStop shoud toggle state.running", () => {
        const startBtn = wrapper.find("button.start");
        startBtn.simulate("click");
        expect(wrapper.state("running")).toBe(true);
        startBtn.simulate("click");
        expect(wrapper.state("running")).toBe(false);
      });
    });
  });

  describe("Should contain lifecycle methods to avoid memory leaks", () => {
    jest.useFakeTimers();
    test("clearInterval should be called on ComponentWillUnmount", () => {
      wrapper.unmount();
      expect(clearInterval).toHaveBeenCalledTimes(1);
    });
  });
});
