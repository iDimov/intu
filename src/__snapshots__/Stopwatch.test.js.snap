// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  <div
    className="Controls"
>
    <button
        className="start"
        href="#"
        onClick={[Function]}
    >
        Start
    </button>
    <button
        href="#"
        onClick={[Function]}
    >
        Lap
    </button>
    <button
        href="#"
        onClick={[Function]}
    >
        Reset
    </button>
    <button
        href="#"
        onClick={[Function]}
    >
        Clear Laps
    </button>
</div>,
  <div
    className="Stopwatch"
>
    00 : 00 : 000
</div>,
  <ul
    className="Results"
/>,
]
`;
